truffle/src/main/java/org/jruby/truffle/core/rope/RopeOperations.java

```diff
@@ -208,6 +208,92 @@ public static LeafRope flatten(Rope rope) {
+         * creating a range, then write flattenBytes in terms of visitBytes.
+         */
+
+        assert length <= rope.byteLength();
+
+        if (rope.getRawBytes() != null) {
+            visitor.accept(rope.getRawBytes(), rope.begin() + offset, length);
+        } else if (rope instanceof ConcatRope) {
+            final ConcatRope concat = (ConcatRope) rope;
+            
+            final int leftLength = concat.getLeft().byteLength();
+
+            if (offset < leftLength) {
+                final int leftUsed;
+
+                if (offset + length > leftLength) {
+                    leftUsed = leftLength - offset;
+                } else {
+                    leftUsed = length;
+                }
+
+                visitBytes(concat.getLeft(), visitor, offset, leftUsed);
+
+                if (leftUsed < length) {
+                    visitBytes(concat.getRight(), visitor, 0, length - leftUsed);
+                }
+            } else {
+                visitBytes(concat.getRight(), visitor, offset - leftLength, length);
+            }
+        } else if (rope instanceof SubstringRope) {
+            final SubstringRope substring = (SubstringRope) rope;
+
+            visitBytes(substring.getChild(), visitor, substring.getOffset() + offset, length);
+        } else if (rope instanceof RepeatingRope) {
+            final RepeatingRope repeating = (RepeatingRope) rope;
+            final Rope child = repeating.getChild();
+
+            final int start = offset % child.byteLength();
+            final int firstPartLength = child.byteLength() - start;
+            visitBytes(child, visitor, start, firstPartLength);
+            final int lengthMinusFirstPart = length - firstPartLength;
+            final int remainingEnd = lengthMinusFirstPart % child.byteLength();
+
+            if (lengthMinusFirstPart >= child.byteLength()) {
+                final byte[] secondPart = child.getBytes();
+
+                final int repeatPartCount = lengthMinusFirstPart / child.byteLength();
+                for (int i = 0; i < repeatPartCount; i++) {
+                    visitBytes(child, visitor, 0, secondPart.length);
+                }
+
+                if (remainingEnd > 0) {
+                    visitBytes(child, visitor, 0, remainingEnd);
+                }
+            } else {
+                visitBytes(child, visitor, 0, remainingEnd);
+            }
+        } else {
+            throw new UnsupportedOperationException("Don't know how to visit rope of type: " + rope.getClass().getName());
+        }
+    }
+
+    @TruffleBoundary
+    public static byte[] extractRange(Rope rope, int offset, int length) {
+        final ByteList result = new ByteList(length);
+
+        visitBytes(rope, new BytesVisitor() {
+
+            @Override
+            public void accept(byte[] bytes, int offset, int length) {
+                result.append(bytes, offset, length);
```

Comment:
`ByteList` has a weird assertion that I think will fail with a small set of `bytes` far enough along in the whole string. It may not matter (and it looks like a bug in `ByteList` to me). But since we run tests with assertions, it might come up. It may be better to just do the byte copying operations on our own here.